File: mac_cleaner/scanners/downloads_browser.py
from __future__ import annotations

import time

from mac_cleaner.models import Category, FileGroup, FileItem
from mac_cleaner.scanners.base import ProgressCallback, Scanner
from mac_cleaner.utils import HOME, list_children, safe_size

INSTALLER_SUFFIXES = {
    ".dmg",
    ".pkg",
    ".zip",
    ".tar",
    ".tar.gz",
    ".tgz",
    ".iso",
    ".app.zip",
}


class DownloadsJunkScanner(Scanner):
    """Finds old installer packages and archives in Downloads."""

    name = "Installer & Download Junk"

    def __init__(self, min_age_days: int = 7, min_size: int = 1024 * 1024):
        self.min_age_days = min_age_days
        self.min_size = min_size

    def scan(self, progress: ProgressCallback | None = None) -> list[FileGroup]:
        self._report(progress, "Scanning Downloads for installers…")
        groups: list[FileGroup] = []
        downloads = HOME / "Downloads"
        if not downloads.exists():
            return groups

        cutoff = time.time() - self.min_age_days * 86400
        for child in list_children(downloads):
            try:
                if not child.is_file():
                    continue
                name_lower = child.name.lower()
                if not any(name_lower.endswith(s) for s in INSTALLER_SUFFIXES):
                    continue
                st = child.stat()
                if st.st_size < self.min_size:
                    continue
                if st.st_mtime > cutoff:
                    continue
            except OSError:
                continue

            age_days = int((time.time() - st.st_mtime) / 86400)
            item = FileItem(
                path=child,
                size=st.st_size,
                category=Category.DOWNLOADS,
                reason=f"Installer/archive unused for {age_days} days",
                group_key=child.name,
            )
            groups.append(
                FileGroup(
                    key=f"dl:{child.name}",
                    category=Category.DOWNLOADS,
                    title=child.name,
                    description=f"~/Downloads — {age_days} days old",
                    items=[item],
                )
            )
        return groups


class BrowserCacheScanner(Scanner):
    name = "Browser Caches"

    BROWSER_PATHS = [
        ("Safari", HOME / "Library" / "Caches" / "com.apple.Safari"),
        ("Chrome", HOME / "Library" / "Caches" / "Google" / "Chrome"),
        ("Chrome", HOME / "Library" / "Application Support" / "Google" / "Chrome" / "Default" / "Code Cache"),
        ("Chrome", HOME / "Library" / "Application Support" / "Google" / "Chrome" / "Default" / "Service Worker" / "CacheStorage"),
        ("Firefox", HOME / "Library" / "Caches" / "Firefox"),
        ("Edge", HOME / "Library" / "Caches" / "com.microsoft.edgemac"),
        ("Brave", HOME / "Library" / "Caches" / "BraveSoftware" / "Brave-Browser"),
        ("Arc", HOME / "Library" / "Caches" / "company.thebrowser.Browser"),
    ]

    def scan(self, progress: ProgressCallback | None = None) -> list[FileGroup]:
        self._report(progress, "Scanning browser caches…")
        groups: list[FileGroup] = []
        seen: set[str] = set()
        for browser, path in self.BROWSER_PATHS:
            try:
                key = str(path.resolve()) if path.exists() else str(path)
            except OSError:
                key = str(path)
            if key in seen:
                continue
            if not path.exists():
                continue
            seen.add(key)
            size = safe_size(path)
            if size < 1024 * 10:
                continue
            item = FileItem(
                path=path,
                size=size,
                category=Category.BROWSER,
                reason=f"{browser} cache",
                group_key=f"{browser}:{path.name}",
            )
            groups.append(
                FileGroup(
                    key=f"browser:{browser}:{path}",
                    category=Category.BROWSER,
                    title=f"{browser} — {path.name}",
                    description=f"{browser} cache data (pages may reload slower once)",
                    items=[item],
                )
            )
        return groups

File: src/scanners/downloads_browser.rs
use std::time::{SystemTime, UNIX_EPOCH};

use crate::models::{Category, FileGroup, FileItem};
use crate::safety::{home_dir, list_children, safe_size};

use super::Scanner;

const INSTALLER_SUFFIXES: &[&str] = &[
    ".dmg", ".pkg", ".zip", ".tar", ".tar.gz", ".tgz", ".iso", ".app.zip",
];

pub struct DownloadsJunkScanner {
    min_age_days: u64,
    min_size: u64,
}

impl Default for DownloadsJunkScanner {
    fn default() -> Self {
        Self {
            min_age_days: 7,
            min_size: 1024 * 1024,
        }
    }
}

impl Scanner for DownloadsJunkScanner {
    fn name(&self) -> &'static str {
        "Installer & Download Junk"
    }

    fn scan(&self, progress: &mut dyn FnMut(&str)) -> Vec<FileGroup> {
        progress("Scanning Downloads for installers…");
        let downloads = home_dir().join("Downloads");
        if !downloads.exists() {
            return Vec::new();
        }

        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_secs())
            .unwrap_or(0);
        let cutoff = now.saturating_sub(self.min_age_days * 86400);

        list_children(&downloads)
            .into_iter()
            .filter(|child| child.is_file())
            .filter_map(|child| {
                let name = child.file_name()?.to_string_lossy().to_lowercase();
                if !INSTALLER_SUFFIXES
                    .iter()
                    .any(|suffix| name.ends_with(suffix))
                {
                    return None;
                }
                let meta = child.metadata().ok()?;
                if meta.len() < self.min_size {
                    return None;
                }
                let mtime = meta
                    .modified()
                    .ok()
                    .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
                    .map(|d| d.as_secs())
                    .unwrap_or(now);
                if mtime > cutoff {
                    return None;
                }
                let age_days = now.saturating_sub(mtime) / 86400;
                let title = child
                    .file_name()
                    .map(|n| n.to_string_lossy().into_owned())
                    .unwrap_or_default();
                Some(FileGroup {
                    key: format!("dl:{title}"),
                    category: Category::Downloads,
                    title: title.clone(),
                    description: format!("~/Downloads — {age_days} days old"),
                    items: vec![FileItem::file(
                        child,
                        meta.len(),
                        Category::Downloads,
                        format!("Installer/archive unused for {age_days} days"),
                        title,
                    )],
                })
            })
            .collect()
    }
}

pub struct BrowserCacheScanner;

impl Scanner for BrowserCacheScanner {
    fn name(&self) -> &'static str {
        "Browser Caches"
    }

    fn scan(&self, progress: &mut dyn FnMut(&str)) -> Vec<FileGroup> {
        progress("Scanning browser caches…");
        let home = home_dir();
        let library = home.join("Library");
        let paths = [
            ("Safari", library.join("Caches").join("com.apple.Safari")),
            (
                "Chrome",
                library.join("Caches").join("Google").join("Chrome"),
            ),
            (
                "Chrome",
                library
                    .join("Application Support")
                    .join("Google")
                    .join("Chrome")
                    .join("Default")
                    .join("Code Cache"),
            ),
            (
                "Chrome",
                library
                    .join("Application Support")
                    .join("Google")
                    .join("Chrome")
                    .join("Default")
                    .join("Service Worker")
                    .join("CacheStorage"),
            ),
            ("Firefox", library.join("Caches").join("Firefox")),
            ("Edge", library.join("Caches").join("com.microsoft.edgemac")),
            (
                "Brave",
                library
                    .join("Caches")
                    .join("BraveSoftware")
                    .join("Brave-Browser"),
            ),
            (
                "Arc",
                library.join("Caches").join("company.thebrowser.Browser"),
            ),
        ];

        let mut seen = std::collections::HashSet::new();
        let mut groups = Vec::new();
        for (browser, path) in paths {
            let key = if path.exists() {
                path.canonicalize().unwrap_or_else(|_| path.clone())
            } else {
                path.clone()
            };
            if !seen.insert(key) || !path.exists() {
                continue;
            }
            let size = safe_size(&path);
            if size < 1024 * 10 {
                continue;
            }
            let leaf = path
                .file_name()
                .map(|n| n.to_string_lossy().into_owned())
                .unwrap_or_default();
            groups.push(FileGroup {
                key: format!("browser:{browser}:{}", path.display()),
                category: Category::Browser,
                title: format!("{browser} — {leaf}"),
                description: format!("{browser} cache data (pages may reload slower once)"),
                items: vec![FileItem::file(
                    path,
                    size,
                    Category::Browser,
                    format!("{browser} cache"),
                    format!("{browser}:{leaf}"),
                )],
            });
        }
        groups
    }
}
